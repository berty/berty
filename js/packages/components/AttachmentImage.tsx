import React, { useState, useEffect } from 'react'
import { Image, ImageProps, ActivityIndicator, View, TouchableOpacity } from 'react-native'
import { useNavigation } from '@react-navigation/native'

import { useMessengerContext } from '@berty-tech/store'

import { getSource } from './utils'

const AttachmentImage: React.FC<{ cid: string; pressable?: boolean } & Omit<ImageProps, 'source'>> =
	props => {
		const { navigate } = useNavigation()
		const { protocolClient, medias } = useMessengerContext()
		const [source, setSource] = useState('')
		const { cid, ...imageProps } = props
		const mimeType = medias[cid]?.mimeType || 'image/jpeg'

		useEffect(() => {
			if (!protocolClient) {
				return
			}
			let cancel = false
			getSource(protocolClient, cid)
				.then(src => {
					if (!cancel) {
						setSource(`data:${mimeType};base64,${src}`)
					}
				})
				.catch(e => console.error('failed to get attachment image:', e))
			return () => {
				cancel = true
			}
		}, [protocolClient, cid, mimeType])

		if (!source) {
			return (
				<View style={[props.style, { alignItems: 'center', justifyContent: 'center' }]}>
					<ActivityIndicator color='grey' />
				</View>
			)
		}

		return !props.pressable ? (
			<Image source={{ uri: source }} {...imageProps} />
		) : (
			<TouchableOpacity
				onPress={() => {
					navigate('Modals.ImageView', { images: [{ uri: source }], previewOnly: true })
				}}
			>
				<Image source={{ uri: source }} {...imageProps} />
			</TouchableOpacity>
		)
	}

export default AttachmentImage
