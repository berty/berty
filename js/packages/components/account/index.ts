export { StreamProgress } from './StreamProgress'
export { CenteredTextScreen } from './CenteredTextScreen'
