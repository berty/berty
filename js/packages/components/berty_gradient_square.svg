<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="171" height="201" version="1.1" viewBox="0 0 171 201"><title>berty_gradient_square</title><defs><linearGradient id="linearGradient-1" x1="50%" x2="50%" y1="0%" y2="100%"><stop offset="0%" stop-color="#4E53FF"/><stop offset="100%" stop-color="#3442D9"/></linearGradient></defs><g id="berty" fill="none" fill-rule="evenodd" stroke="none" stroke-width="1"><g id="onboarding/0" fill-rule="nonzero" transform="translate(-100.000000, -183.000000)"><g id="berty_gradient_square" transform="translate(100.000000, 183.000000)"><g id="Group" fill="#F74868" transform="translate(0.000000, 128.000000)"><path id="Shape" d="M15.0310093,18.7802133 C13.1974624,18.7802133 11.8243396,19.0893273 10.8342243,19.4700903 L10.8342243,0 L0,0 L0,51.9802895 C3.05387407,53.8963869 7.85980396,55.6609984 13.7332656,55.6609984 C25.5616797,55.6609984 33.9532123,48.6066476 33.9532123,36.6453671 C33.951175,25.9880976 25.1012558,18.7802133 15.0310093,18.7802133 Z M14.4972434,45.9208354 C12.971325,45.9208354 11.7509978,45.6158155 10.8342243,45.2309586 L10.8342243,29.4395299 C11.6735813,28.9031863 13.4276745,28.4405388 14.9535928,28.4405388 C19.2277942,28.4405388 23.1943671,31.8919711 23.1943671,37.0281772 C23.1943671,42.319964 19.8369391,45.9208354 14.4972434,45.9208354 Z"/><path id="Path" d="M88.8074831,25.2691917 L88.5042817,19.8217224 L78.2157821,19.8217224 L78.2157821,54.5798135 L89.3406294,54.5798135 L89.3406294,40.8453893 C89.3406294,32.4826326 94.826744,28.9527381 100.770713,28.9527381 C101.303859,28.9527381 101.533804,28.9527381 101.686422,28.9527381 L101.780028,19.9712769 C96.1392602,19.9692282 90.9421029,20.8952365 88.8074831,25.2691917 Z"/><path id="Path" d="M121.496438,9.00987383 L110.835353,13.0751966 L110.835353,19.8234276 L106.755587,19.8234276 L106.828803,29.0292845 L110.605536,29.0292845 L110.605536,43.2179778 C110.605536,50.1976355 115.783544,55.1804717 123.400057,55.1804717 C125.610778,55.1804717 127.514398,54.8753165 128.65535,54.3366871 L129.493268,45.1349261 C128.50485,45.5199541 127.361864,45.6715078 125.913812,45.6715078 C123.398024,45.6715078 121.648971,44.2133161 121.648971,41.606185 L121.648971,29.0292845 L131.472138,29.0292845 L132.310056,19.8234276 L121.494404,19.8234276 L121.494404,9.00987383 L121.496438,9.00987383 Z"/><path id="Path" d="M160.452905,19.8237714 L155.257915,33.4820282 C154.492746,35.55351 153.805114,37.7028517 153.195019,40.1574858 L153.039944,40.1574858 C152.352312,37.6270408 151.66468,35.3998393 150.901551,33.4062173 L145.780017,19.8217224 L134.698324,19.8217224 L147.612343,52.8179074 C144.327215,60.8005908 140.966592,63.9457191 135.616527,65.7119083 L140.201422,73 C148.530546,70.9285183 154.186678,63.7940973 158.84707,51.59059 L171,19.8217224 L160.452905,19.8217224 L160.452905,19.8237714 Z"/><path id="Shape" d="M56.4471314,18.7405376 C46.7640176,18.7405376 38.4511173,25.7249911 38.4511173,37.0091464 C38.4511173,48.367081 47.3727752,55.6609984 58.658204,55.6609984 C63.2330479,55.6609984 66.5903762,54.8166338 68.9521115,53.6648548 L69.9456688,44.376843 C66.9711043,46.1414015 63.9191727,47.0636442 59.9551224,47.0636442 C53.8553307,47.0636442 50.6507007,43.5324784 49.8139133,40.4603845 L64.7722473,40.5034225 C64.7722473,40.5034225 70.9982698,40.5034225 71.6884665,35.9782837 C71.6884665,35.9782837 71.926676,33.6870222 71.926676,32.5577869 C71.926676,25.4196262 66.0549141,18.7405376 56.4471314,18.7405376 Z M61.6327685,33.2504939 L49.4331855,33.2504939 C50.1966774,29.7193279 53.0938748,27.2641122 56.4491674,27.2641122 C59.6537974,27.2641122 61.7101357,29.7193279 61.7101357,32.481958 C61.7101357,32.7873229 61.7101357,33.0189084 61.6327685,33.2504939 Z"/></g><g id="Group" transform="translate(22.000000, 0.000000)"><g id="Path"><path fill="url(#linearGradient-1)" d="M54.0087319,33.8157583 C54.0087319,33.409156 54.0166744,33.0025537 54.0365305,32.5999377 C54.6679561,18.5362815 66.5220785,7.33080037 80.6807107,7.29891 C80.6953994,4.67194021 80.4015284,2.96580505 80.4015284,2.90601059 C80.4531544,2.90601059 80.5047804,2.90202429 80.5524352,2.90202429 C72.5980609,2.82628465 64.7389961,4.4088446 57.4676735,7.47032077 C55.0055107,8.50675802 52.6148301,9.71061974 50.3075453,11.0779197 L50.3075453,11.0779197 C47.4085472,12.7960137 44.6485421,14.7692308 42.0632712,16.9975709 C41.2015142,17.7390221 40.3596134,18.5083775 39.5375687,19.3056369 L25.0942042,11.2812208 L17.1358587,6.86440361 L4.77341881,0.0039862971 C-5.36116089,18.30109 1.18739786,41.3737776 19.3915175,51.5667393 L1.50112505,85.856867 C1.50509628,85.856867 1.50509628,85.8608533 1.50906751,85.8608533 C9.05043383,89.8232326 17.0643765,91.9439427 25.0425782,92.4103395 C47.146446,93.705886 69.0040975,82.2094052 79.9448369,61.2414824 L80.3657873,60.336593 C80.2347367,60.336593 80.0997149,60.336593 79.9686643,60.3326067 C65.5928107,60.0854562 54.0087319,48.3099346 54.0087319,33.8157583 Z" transform="translate(10.126637, 17.101215)"/><path fill="#FFBF47" d="M60.8869028,49.2307692 C56.8680177,44.7980069 54.3224591,38.9939583 54.0365305,32.5999377 C54.0166744,33.0025537 54.0087319,33.409156 54.0087319,33.8157583 C54.0087319,48.3099346 65.5888394,60.0894425 79.9726355,60.3326067 C80.1275135,60.336593 80.2784203,60.336593 80.4332982,60.336593 L80.4332982,57.9009654 C72.6893992,57.9049517 65.71989,54.5604485 60.8869028,49.2307692 Z" transform="translate(10.126637, 17.101215)"/><path fill="#FFF" d="M80.3697586,7.2949237 C66.2123226,7.32681408 54.6719274,18.5362815 54.0365305,32.5999377 C54.3224591,38.9939583 56.8680177,44.7980069 60.8869028,49.2307692 C65.71989,54.5604485 72.685428,57.9049517 80.4332982,57.9049517 L80.4332982,7.2909374 C80.4134421,7.2909374 80.3935859,7.2909374 80.3697586,7.2949237 Z" transform="translate(10.126637, 17.101215)"/></g><g><path id="Path" fill="#FFBF47" d="M14.900056,17.1052009 L6.50487521,12.4173155 C-6.2149754,35.3823731 2.16829174,64.4743694 25.0465494,77.242479 L29.5658095,68.695858 C11.3219776,58.5148552 4.75753388,35.4182498 14.900056,17.1052009 Z"/><g id="Path"><path fill="#FFAE3A" d="M0.0238273817,36.6699471 L0.0238273817,57.4545002 C0.0317698423,57.4545002 0.0436835332,57.4545002 0.0516259937,57.4545002 C9.99558663,57.4425413 18.2954579,50.3907817 20.2771018,40.9990657 L0.0238273817,36.6699471 Z" transform="translate(90.504338, 19.971348)"/><path fill="#FFBF47" d="M25.0147796,9.60697602 C19.0221931,4.14574899 11.226668,0.649766428 2.63292568,0.107630022 C1.81485224,0.0637807537 0.996778802,0.0358766739 0.178705363,0.0279040797 C0.127079369,0.0279040797 0.0754533755,0.0318903768 0.027798612,0.0318903768 C0.027798612,0.167424478 0.027798612,36.6699471 0.027798612,36.6699471 L36.316901,44.4193086 C36.8133048,42.0036126 37.0793772,39.5042043 37.0793772,36.9410153 C37.0714347,26.1062597 32.4211241,16.3597633 25.0147796,9.60697602 Z" transform="translate(90.504338, 19.971348)"/></g><g id="Path"><g id="Group" transform="translate(27.242640, 0.000000)"><path id="Path" fill="#3442D9" d="M33.1875715,28.1751479 L24.1569938,8.06826534 C15.503683,11.9867954 9.75334159,19.7122392 7.97820165,28.3824354 L22.4255374,36.3988789 C25.7057737,33.2138275 29.3275357,30.4632825 33.1875715,28.1751479 Z"/><path id="Path" fill="#FFBF47" d="M7.97820165,28.3824354 C9.75731282,19.7122392 15.503683,11.9867954 24.1569938,8.06826534 L20.5471455,0.0318903768 C10.1107523,4.75565244 2.90694057,13.7088757 0.0158849212,23.9616319 L7.97820165,28.3824354 L7.97820165,28.3824354 Z"/></g><path fill="#F74868" d="M0.027798612,125.161756 C12.0209141,131.480037 26.8455167,126.843974 33.1399168,114.805357 C34.0374148,113.087263 34.7085527,111.31336 35.1692154,109.511554 C27.1870425,109.045157 19.1770711,106.924447 11.6357047,102.962068 L0.027798612,125.161756 Z"/></g><ellipse id="Oval" cx="82.054" cy="49.259" fill="#383B62" rx="4.734" ry="4.752"/></g></g></g></g></g></svg>