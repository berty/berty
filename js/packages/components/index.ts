export * from './buttons'
export * from './create-group'
export * from './inputs'
