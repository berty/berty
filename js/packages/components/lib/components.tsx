import LibExample from "./Example.tsx"

export const Example = LibExample
