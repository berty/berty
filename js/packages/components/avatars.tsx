import React from 'react'
import { Image, View, ViewStyle, Text, TouchableOpacity } from 'react-native'
import palette from 'google-palette'
import { SHA3 } from 'sha3'
import { Buffer } from 'buffer'
import { withBadge } from 'react-native-elements'

import { useStyles } from '@berty-tech/styles'
import {
	useAccount,
	useContact,
	useMember,
	useConversation,
	Maybe,
	useMsgrContext,
} from '@berty-tech/store/hooks'
import { navigate } from '@berty-tech/navigation'
import beapi from '@berty-tech/api'
import PinkBotAvatar from '@berty-tech/assets/berty_bot_pink_bg.png'
import GreenDevAvatar from '@berty-tech/assets/berty_dev_green_bg.png'
import OrangeBotAvatar from '@berty-tech/assets/berty_bot_orange_bg.png'
import BlueDevAvatar from '@berty-tech/assets/berty_dev_blue_bg.png'

import AttachmentImage from './AttachmentImage'
import GroupAvatar from './main/Avatar_Group_Copy_19.png'
import Logo from './main/1_berty_picto.svg'

export type AvatarStyle = Omit<
	ViewStyle,
	'borderRadius' | 'width' | 'height' | 'alignItems' | 'justifyContent'
>

const pal = palette('tol-rainbow', 256)

export const GenericAvatar: React.FC<{
	cid: Maybe<string>
	colorSeed: Maybe<string>
	size: number
	style?: AvatarStyle
	isEditable?: boolean
	nameSeed: Maybe<string>
	pressable?: boolean
}> = ({ cid, size, colorSeed, style, isEditable = false, nameSeed, pressable }) => {
	const [{ border, background, color }] = useStyles()
	const padding = Math.round(size / 14)
	let innerSize = Math.round(size - 2 * padding)
	let content: JSX.Element
	if (cid) {
		if (innerSize % 2) {
			innerSize--
		}
		content = (
			<>
				<AttachmentImage
					cid={cid}
					style={{ width: innerSize, height: innerSize, borderRadius: innerSize / 2 }}
					pressable={pressable}
				/>
				{isEditable ? (
					<View
						style={[
							{
								width: innerSize,
								height: innerSize,
								position: 'absolute',
								backgroundColor: color.light.blue,
								opacity: 0.6,
							},
							border.radius.scale(innerSize / 2),
						]}
					/>
				) : null}
			</>
		)
	} else {
		let iconSize = Math.round(innerSize - innerSize / 10) // adjust for jdenticon bug
		if (iconSize % 2) {
			iconSize--
		}
		content = (
			<>
				<NameAvatar size={size} style={style} colorSeed={colorSeed} nameSeed={nameSeed} />
				{isEditable ? (
					<View
						style={[
							{
								width: innerSize,
								height: innerSize,
								position: 'absolute',
								backgroundColor: color.light.blue,
								opacity: 0.6,
							},
							border.radius.scale(innerSize / 2),
						]}
					/>
				) : null}
			</>
		)
	}
	return (
		<>
			<View
				style={[
					background.white,
					border.shadow.medium,
					style,
					{
						borderRadius: size / 2,
						width: size,
						height: size,
						alignItems: 'center',
						justifyContent: 'center',
					},
				]}
			>
				{content}
			</View>
		</>
	)
}

const hardcodedAvatars = {
	berty_dev_green_bg: GreenDevAvatar,
	berty_bot_pink_bg: PinkBotAvatar,
	berty_dev_blue_bg: BlueDevAvatar,
	berty_bot_orange_bg: OrangeBotAvatar,

	group: GroupAvatar,
}

export type HardcodedAvatarKey = keyof typeof hardcodedAvatars

export const HardcodedAvatar: React.FC<{
	size: number
	style?: AvatarStyle
	name: HardcodedAvatarKey
	pressable?: boolean
}> = ({ size, style, name, pressable }) => {
	const [{ border }] = useStyles()
	let avatar = hardcodedAvatars[name]
	if (!avatar) {
		avatar = Logo
	}

	return (
		<TouchableOpacity
			activeOpacity={0.9}
			disabled={!pressable}
			onPress={() => {
				navigate('ImageView', { images: [avatar], previewOnly: true })
			}}
			style={[{ borderRadius: size / 2, backgroundColor: 'white' }, border.shadow.medium, style]}
		>
			<Image
				source={avatar}
				style={{
					width: size,
					height: size,
					borderRadius: size / 2,
				}}
			/>
		</TouchableOpacity>
	)
}

export const AccountAvatar: React.FC<{
	size: number
	style?: AvatarStyle
	isEditable?: boolean
}> = ({ size, style, isEditable }) => {
	const account = useAccount()
	return (
		<GenericAvatar
			nameSeed={account?.displayName}
			cid={account?.avatarCid}
			size={size}
			colorSeed={account?.publicKey}
			style={style}
			isEditable={isEditable}
		/>
	)
}

export const NameAvatar: React.FC<{
	colorSeed: Maybe<string>
	size: number
	style?: AvatarStyle
	nameSeed: Maybe<string>
}> = ({ colorSeed, size, style, nameSeed }) => {
	const h = new SHA3(256).update(Buffer.from(colorSeed || '', 'base64')).digest()
	const color = '#' + pal[h[0]]
	return (
		<View
			style={[
				style,
				{
					width: size,
					height: size,
					backgroundColor: color,
					borderRadius: size / 2,
					alignItems: 'center',
					justifyContent: 'center',
				},
			]}
		>
			<Text style={{ color: 'white', fontSize: size * 0.5, includeFontPadding: false }}>
				{(nameSeed || '?')[0].toUpperCase()}
			</Text>
		</View>
	)
}

export const ContactAvatar: React.FC<{
	publicKey: Maybe<string>
	size: number
	style?: AvatarStyle
	fallbackNameSeed?: Maybe<string>
	pressable?: boolean
}> = ({ publicKey, size, style, fallbackNameSeed, pressable }) => {
	const contact = useContact(publicKey)
	const ctx = useMsgrContext()
	const suggestion = Object.values(ctx.persistentOptions?.suggestions).find(
		(v) => v.pk === publicKey,
	)
	if (suggestion) {
		return (
			<HardcodedAvatar
				size={size}
				style={style}
				name={suggestion.icon as any}
				pressable={pressable}
			/>
		)
	}
	return (
		<GenericAvatar
			nameSeed={contact?.displayName || fallbackNameSeed}
			cid={contact?.avatarCid}
			size={size}
			colorSeed={publicKey}
			style={style}
			pressable={pressable}
		/>
	)
}

export const MemberAvatar: React.FC<{
	publicKey: Maybe<string>
	conversationPublicKey: Maybe<string>
	size: number
	pressable?: boolean
}> = ({ publicKey, conversationPublicKey, size, pressable }) => {
	const member = useMember({ publicKey, conversationPublicKey })
	return (
		<GenericAvatar
			cid={member?.avatarCid}
			size={size}
			colorSeed={publicKey}
			nameSeed={member?.displayName}
			pressable={pressable}
		/>
	)
}

export const MultiMemberAvatar: React.FC<{
	size: number
	style?: AvatarStyle
	publicKey?: Maybe<string>
	fallbackNameSeed?: Maybe<string>
	pressable?: boolean
}> = ({ size, style, publicKey, fallbackNameSeed, pressable }) => {
	const ctx = useMsgrContext()
	const conv = useConversation(publicKey)
	const suggestion = Object.values(ctx.persistentOptions?.suggestions).find(
		(v) => v.pk === publicKey,
	)
	let content: React.ReactElement
	if (suggestion) {
		content = (
			<HardcodedAvatar
				size={size}
				style={style}
				name={suggestion.icon as any}
				pressable={pressable}
			/>
		)
	} else {
		content = (
			<GenericAvatar
				size={size}
				style={style}
				cid={conv?.avatarCid}
				colorSeed={publicKey}
				nameSeed={conv?.displayName || fallbackNameSeed}
				pressable={pressable}
			/>
		)
	}
	const badgeSize = size / 3
	class GroupBadge extends React.Component {
		render = () => <HardcodedAvatar size={badgeSize} name={'group'} />
	}
	const Avatar = () => content
	const WrappedAvatar = withBadge('', { Component: GroupBadge })(Avatar)
	return <WrappedAvatar />
}

export const ConversationAvatar: React.FC<{
	publicKey: Maybe<string>
	size: number
	style?: AvatarStyle
}> = ({ publicKey, size, style }) => {
	const conv = useConversation(publicKey)
	const ctx = useMsgrContext()

	if (conv) {
		if (conv.type === beapi.messenger.Conversation.Type.MultiMemberType) {
			return <MultiMemberAvatar size={size} style={style} publicKey={publicKey} />
		} else if (conv.type === beapi.messenger.Conversation.Type.ContactType) {
			return <ContactAvatar size={size} publicKey={conv?.contactPublicKey} />
		}
	}

	const suggestion = Object.values(ctx.persistentOptions?.suggestions).find(
		(v) => v.pk === publicKey,
	)
	if (suggestion) {
		return <HardcodedAvatar size={size} style={style} name={suggestion.icon as any} />
	}

	return <GenericAvatar size={size} style={style} cid='' colorSeed={publicKey} nameSeed={'C'} />
}
