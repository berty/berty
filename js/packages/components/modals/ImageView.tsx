import React, { useState } from 'react'
import { View, Modal, TouchableOpacity, Image } from 'react-native'
import { Text, Icon } from '@ui-kitten/components'
import { useTranslation } from 'react-i18next'
import CameraRoll from '@react-native-community/cameraroll'
import Share from '../../rnutil/react-native-share'
import ImageViewer from 'react-native-image-zoom-viewer'

import { useStyles } from '@berty-tech/styles'
import { useThemeColor } from '@berty-tech/store'
import { ScreenFC, useNavigation } from '@berty-tech/navigation'

import { ForwardToBertyContactModal } from './ForwardToBertyContactModal'

export const ImageView: ScreenFC<'Modals.ImageView'> = ({
	route: {
		params: { images, previewOnly = false },
	},
}) => {
	const [{ border, padding }] = useStyles()
	const colors = useThemeColor()
	const { t }: { t: any } = useTranslation()
	const { goBack } = useNavigation()

	const [currentIndex, setCurrentIndex] = useState(0)
	const [isModalVisible, setModalVisibility] = useState(false)
	const [isForwardModalVisible, setForwardModalVisibility] = useState(false)
	const [message, setMessage] = useState('')

	const handleMessage = (msg: string) => {
		setMessage(msg)
		setTimeout(() => setMessage(''), 1400)
	}

	const MENU_LIST = [
		{
			title: t('chat.files.save-to-gallery'),
			onPress() {
				const uri = images[currentIndex]?.uri
				if (!uri) {
					return
				}
				CameraRoll.save(uri, { type: 'photo' })
					.then(() => {
						setModalVisibility(false)
						handleMessage(t('chat.files.image-saved'))
					})
					.catch(err => console.log(err))
			},
		},
		{
			title: t('chat.files.share'),
			onPress() {
				const img = images[currentIndex]
				if (!img) {
					return
				}
				Share.open({
					title: img.displayName || img.filename || 'Image from Berty',
					url: img.uri,
				})
					.then(() => {})
					.catch(err => {
						err && console.log(err)
					})
			},
		},
		{
			title: t('chat.files.forward-berty'),
			onPress() {
				setForwardModalVisibility(true)
				setModalVisibility(false)
			},
		},
	]

	return (
		<Modal transparent>
			<ImageViewer
				imageUrls={images.map(image => ({
					url: image.uri || Image.resolveAssetSource(image).uri,
				}))}
				index={0}
				onClick={() => {
					setModalVisibility(prev => !prev)
				}}
				onChange={index => {
					index && setCurrentIndex(index)
				}}
				renderFooter={() => <></>}
				renderIndicator={previewOnly ? () => <></> : undefined}
				enablePreload
				enableSwipeDown
				onSwipeDown={goBack}
			/>

			{isModalVisible && (
				<Modal transparent animationType='fade'>
					<TouchableOpacity
						style={[padding.medium, { position: 'absolute', top: 50, left: 10, zIndex: 9 }]}
						activeOpacity={0.8}
						onPress={goBack}
					>
						<Icon
							name='arrow-back-outline'
							fill={colors['reverted-main-text']}
							style={{ opacity: 0.8 }}
							height={30}
							width={30}
						/>
					</TouchableOpacity>
					<TouchableOpacity onPress={() => setModalVisibility(false)} style={{ flex: 1 }} />
					{!previewOnly && (
						<View
							style={[
								{
									position: 'absolute',
									left: 0,
									bottom: 0,
									right: 0,
									backgroundColor: colors['main-background'],
								},
								padding.medium,
								border.radius.top.large,
							]}
						>
							{MENU_LIST.map(item => (
								<TouchableOpacity key={item.title} onPress={item.onPress} style={[padding.medium]}>
									<Text style={{ textAlign: 'center' }}>{item.title}</Text>
								</TouchableOpacity>
							))}
						</View>
					)}
				</Modal>
			)}
			{!!message && (
				<View
					style={[
						{
							position: 'absolute',
							bottom: 100,
							alignItems: 'center',
							justifyContent: 'center',
							right: 0,
							left: 0,
						},
					]}
				>
					<View
						style={[
							border.radius.large,
							padding.vertical.small,
							padding.horizontal.large,
							{ backgroundColor: colors['main-background'] },
						]}
					>
						<Text style={{ color: 'black' }}>{message}</Text>
					</View>
				</View>
			)}
			{isForwardModalVisible && (
				<ForwardToBertyContactModal
					image={images[currentIndex]}
					onClose={() => setForwardModalVisibility(false)}
				/>
			)}
		</Modal>
	)
}
