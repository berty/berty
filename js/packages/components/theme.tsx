import React from 'react'
import { ApplicationProvider } from '@ui-kitten/components'
import { mapping, light } from '@eva-design/eva'
import { StatusBarWrapper } from './common'

export const Provider: React.FC = ({ children }) => (
	<>
		<ApplicationProvider
			mapping={mapping}
			customMapping={
				{
					strict: {
						'text-font-family': 'Open Sans',
						'text-paragraph-1-font-weight': '600',
					},
				} as any
			}
			theme={light}
		>
			<StatusBarWrapper />
			{children}
		</ApplicationProvider>
	</>
)
