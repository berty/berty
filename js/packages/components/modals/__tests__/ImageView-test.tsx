import React from 'react'
import renderer from 'react-test-renderer'
import { NavigationContainer } from '@react-navigation/native'
import { IconRegistry } from '@ui-kitten/components'
import { EvaIconsPack } from '@ui-kitten/eva-icons'
import { SafeAreaProvider } from 'react-native-safe-area-context'

import { Provider as ThemeProvider } from '@berty-tech/components/theme'
import { FeatherIconsPack } from '@berty-tech/messenger-app/feather-icons'
import { CustomIconsPack } from '@berty-tech/messenger-app/custom-icons'
import { Provider as StyleProvider } from '@berty-tech/styles'

import { ImageView } from '../ImageView'

jest.mock('react-native/Libraries/LogBox/LogBox')

describe('ImageView modal', () => {
	beforeEach(() => {
		jest.resetModules()
		jest.resetAllMocks()
		jest.useFakeTimers()
	})
	it('renders correctly', () => {
		const TestProvider: React.FC = ({ children }) => (
			<SafeAreaProvider
				initialMetrics={{
					frame: { x: 0, y: 0, width: 0, height: 0 },
					insets: { top: 0, left: 0, right: 0, bottom: 0 },
				}}
			>
				<StyleProvider>
					<IconRegistry icons={[EvaIconsPack, FeatherIconsPack, CustomIconsPack]} />
					<ThemeProvider>
						<NavigationContainer>{children}</NavigationContainer>
					</ThemeProvider>
				</StyleProvider>
			</SafeAreaProvider>
		)
		// FIXME: add Media fixture
		const props: React.ComponentProps<typeof ImageView> = {
			route: {
				params: {
					images: [],
				},
			},
		}
		const tree = renderer
			.create(
				<TestProvider>
					<ImageView {...props} />
				</TestProvider>,
			)
			.toJSON()
		expect(tree).toMatchSnapshot()
	})
})
