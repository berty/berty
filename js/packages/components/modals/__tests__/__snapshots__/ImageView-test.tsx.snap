// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageView modal renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Modal
      hardwareAccelerated={false}
      transparent={true}
      visible={true}
    >
      <View
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
            "overflow": "hidden",
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "zIndex": 9,
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "transparent",
                  "opacity": 0,
                }
              }
            >
              <View
                style={
                  Object {
                    "bottom": 0,
                    "justifyContent": "center",
                    "left": 0,
                    "position": "absolute",
                    "top": 0,
                    "zIndex": 13,
                  }
                }
              >
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                />
              </View>
              <View
                style={
                  Object {
                    "bottom": 0,
                    "justifyContent": "center",
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "zIndex": 13,
                  }
                }
              >
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                />
              </View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                    ],
                    "width": 0,
                  }
                }
              />
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 38,
                    "zIndex": 13,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "color": "white",
                      "fontSize": 16,
                      "textShadowColor": "rgba(0, 0, 0, 0.3)",
                      "textShadowOffset": Object {
                        "height": 0.5,
                        "width": 0,
                      },
                      "textShadowRadius": 0,
                    }
                  }
                >
                  1/0
                </Text>
              </View>
              <View
                style={
                  Array [
                    Object {
                      "bottom": 0,
                      "position": "absolute",
                      "zIndex": 9,
                    },
                    Object {},
                  ]
                }
              />
            </View>
          </View>
        </View>
      </View>
    </Modal>
  </View>
</RNCSafeAreaProvider>
`;
